### PR TITLE
#10524 - SetpointManager:ReturnTemperature:Chilled/HotWater (useless) warnings from GetSpecificHeatGlycol 

### LINKED ISSUE #10524: SetpointManager:ReturnTemperature:Chilled/HotWater never reads PlantLoop fluid glycol (E+ Bugfix)
Issue overview
--------------
This is based on my post on UnmetHours: https://unmethours.com/question/99091/setpointmanagerreturntemperaturechilledhotwater-never-reads-plantloop-fluid-glycol-e-bugfix/

I work with E+9.4 but this bug is present on the most recent version. However I don't know how to modify the SC there so I can only explain it in E+ 9.4 which I believe is more than enough in this case, the issue is relatively simple.

The short version is that the  SetpointManager:ReturnTemperature:Chilled/HotWater object has a bug that makes it not read the fluid of the plant that it is in and always uses water. Therefore if using any glycol the heat balance is wrong in the function that regulates the impulsion temperature.

The only affected file is Setpointmanager.cc. I will use the lines for the subroutine of the chilledwater for the explanation but the hotwater has the exact same code and bug. The subroutine that doesn't work properly is "DefineReturnWaterChWSetPointManager" and its hotwater counterpart, for chillwater that is lines 7606 to 7624:

```
// we need to know the plant to get the fluid ID in case it is glycol
   // but we have to wait in case plant isn't initialized yet
   // if plant isn't initialized, assume index=1 (water)
   int fluidIndex = 1;
   **if (this->plantLoopIndex == 0) {** **This here is the problem, this only happens once, at first iteration**
       for (int plantIndex = 1; plantIndex <= DataPlant::TotNumLoops; plantIndex++) {
           if (this->supplyNodeIndex == DataPlant::PlantLoop(plantIndex).LoopSide(2).NodeNumOut) {
               this->plantLoopIndex = plantIndex;
               this->plantSetpointNodeIndex = DataPlant::PlantLoop(plantIndex).TempSetPointNodeNum;
               fluidIndex = DataPlant::PlantLoop(plantIndex).FluidIndex; **So this is correctly assigned once**
               // now that we've found the plant populated, let's verify that the nodes match
               if (!PlantUtilities::verifyTwoNodeNumsOnSamePlantLoop(this->supplyNodeIndex, this->returnNodeIndex)) {
                   ShowSevereError("Node problem for SetpointManager:ReturnTemperature:ChilledWater.");
                   ShowContinueError("Return and Supply nodes were not found on the same plant loop.  Verify node names.");
                   ShowFatalError("Simulation aborts due to setpoint node problem");
               }
           }
       }
   }
```

The important thing is that "fluidIndex" is the variable that tells the setpointmanager what the fluid is. It always starts as water and then inside that for loop it is changed to the fluid on the plant. But that loop activates only once, on first iteration. Because after the plandLoopIndex is set to the plantIndex that condition is never entered.

I checked that by puting some "ShowSevereError" inside that loop and it only enters the condition once and only on first iteration, therefore every single call to this subroutine, which occurs every timestep, sets the fluid as water but keeps the plantLoopIndex from the first iteration and it doesn't assignagain the fluidindex that was changed to water.

A simple fix is adding this afterwards: (I did it on my source code for 9.4 and I believe it works fine. The setpoint manager works the same way with water and with glycol it stops giving warnings about the fluid being water and its properties not calculable below 0C, btw that is the warning that led me to this issue, the explanation in the unmethours link):

```
 if (this->plantLoopIndex != 0) {
     fluidIndex = DataPlant::PlantLoop(this->plantLoopIndex).FluidIndex;
 }

```

With that it will initialise the plant, then change the fluid (which is redundant) and then if the plant is initialised in every subsequent timestep it will still change the fluidIndex to the correct fluid of the plant.
Another possible solution is setting this->plantLoopIndex=0 at the start of the subroutine but that just makes the calculations longer. Maybe using and if condition when assigning to water as a default is even better in computation speed.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): Win10
 - Version of EnergyPlus: 9.4
 - Unmethours link or helpdesk ticket number: https://unmethours.com/question/99091/setpointmanagerreturntemperaturechilledhotwater-never-reads-plantloop-fluid-glycol-e-bugfix/

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (Setpointmanager.cc)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #10524 

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/SetPointManager.unit.cc [hunk 1]
```diff
@@ -173,9 +174,81 @@ TEST_F(EnergyPlusFixture, SetPointManager_DefineReturnWaterChWSetPointManager)
...
+    //   **   ~~~   **  Environment=, at Simulation time= 00:00 - 00:00
+    state->dataLoopNodes->Node(1).Temp = 1;
+    state->dataLoopNodes->Node(2).Temp = -3;
+    // Calling it a second time, ensure #10524 is fixed
+    mySPM.calculate(*state, state->dataLoopNodes->Node(1), state->dataLoopNodes->Node(2));
+    // on the first pass through it should detect the plant loop it manages
+    EXPECT_EQ(1, mySPM.plantLoopIndex);
+    // with a delta T of 4, and a target return of 12, it should produce 8
+    EXPECT_EQ(8, mySPM.currentSupplySetPt);
+    compare_err_stream("");
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SetPointManager.unit.cc]: Test that shows the issue originally.

File: src/EnergyPlus/SetPointManager.cc [hunk 2]
```diff
@@ -8282,18 +8279,19 @@ void DefineReturnWaterHWSetPointManager::calculate(EnergyPlusData &state, DataLo
...
     }
 
-    // we don't need fluid names since we have a real index, so just pass in the temperature and get properties
-    Real64 avgTemp = (returnNode.Temp + supplyNode.Temp) / 2;
-    Real64 cp = FluidProperties::GetSpecificHeatGlycol(state, "", avgTemp, fluidIndex, "ReturnWaterHWSetPointManager::calculate");
-
     // get the operating flow rate
-    Real64 mdot = supplyNode.MassFlowRate;
+    Real64 const mdot = supplyNode.MassFlowRate;
+    Real64 const deltaT = supplyNode.Temp - returnNode.Temp;
```
Comment (jmarrec) [inline src/EnergyPlus/SetPointManager.cc]: That's really all we need

File: src/EnergyPlus/SetPointManager.cc [hunk 3]
```diff
@@ -8282,18 +8279,19 @@ void DefineReturnWaterHWSetPointManager::calculate(EnergyPlusData &state, DataLo
...
+    // // calculate the current demand
+    // fluidIndex = state.dataPlnt->PlantLoop(this->plantLoopIndex).FluidIndex;
+    // // we don't need fluid names since we have a real index, so just pass in the temperature and get properties
+    // Real64 const avgTemp = (returnNode.Temp + supplyNode.Temp) / 2;
+    // Real64 const cp = FluidProperties::GetSpecificHeatGlycol(state, "", avgTemp, fluidIndex, "ReturnWaterHWSetPointManager::calculate");
+    // Real64 const Real64 Qdemand = mdot * cp * deltaT;
 
     // check for strange conditions
-    if (Qdemand < 0) {
+    if (deltaT < 0) { // supplyNode.Temp < returnNode.Temp, which shouldn't happen for a HW Loop
```
Comment (jmarrec) [inline src/EnergyPlus/SetPointManager.cc]: Checking the sign of Qdemand can be done by checking the sign deltaT only.

File: src/EnergyPlus/SetPointManager.cc [hunk 4]
```diff
@@ -8321,7 +8319,7 @@ void DefineReturnWaterHWSetPointManager::calculate(EnergyPlusData &state, DataLo
     // calculate the supply setpoint to use, default to the design value if flow is zero
     Real64 T_supply_setpoint = this->maximumHotWaterSetpoint;
     if (mdot > DataConvergParams::PlantFlowRateToler) {
-        T_supply_setpoint = T_return_target + Qdemand / (mdot * cp);
+        T_supply_setpoint = T_return_target + deltaT;
```
Comment (jmarrec) [inline src/EnergyPlus/SetPointManager.cc]: Using `Qdemand / (mdot * cp) `means just using `deltaT`

So we do not need the call to GetSpecificHeatGlycol nor the calculation of Qdemand

File: src/EnergyPlus/SetPointManager.cc [hunk 5]
```diff
@@ -8282,18 +8279,19 @@ void DefineReturnWaterHWSetPointManager::calculate(EnergyPlusData &state, DataLo
...
+    Real64 const deltaT = supplyNode.Temp - returnNode.Temp;
 
-    // calculate the current demand
-    Real64 Qdemand = mdot * cp * (supplyNode.Temp - returnNode.Temp);
+    // // calculate the current demand
+    // fluidIndex = state.dataPlnt->PlantLoop(this->plantLoopIndex).FluidIndex;
+    // // we don't need fluid names since we have a real index, so just pass in the temperature and get properties
+    // Real64 const avgTemp = (returnNode.Temp + supplyNode.Temp) / 2;
+    // Real64 const cp = FluidProperties::GetSpecificHeatGlycol(state, "", avgTemp, fluidIndex, "ReturnWaterHWSetPointManager::calculate");
+    // Real64 const Real64 Qdemand = mdot * cp * deltaT;
```
Comment (jmarrec) [inline src/EnergyPlus/SetPointManager.cc]: Maybe I should just trash this commented out code?

**Comment by Myoldmopar** [comment]
I'm going to hold here.  With @amirroth 's refactor of SPM code, I'm afraid there will be conflicts :) and I'd rather deal with them in this small context PR rather than in that gigantic one.  Please hold and stay on the telephone line, your PR is very important.

**Comment by Myoldmopar** [comment]
@jmarrec this will need the conflict(s) resolved.  I haven't looked, but GitHub is offering to let me resolve them online, so they must not be too bad!

**Comment by jmarrec** [comment]
@Myoldmopar sure, I'll resolve!

**Comment by jmarrec** [comment]
Replayed the changes onto develop manually.

File: tst/EnergyPlus/unit/SetPointManager.unit.cc [hunk 6]
```diff
@@ -1517,7 +1590,8 @@ TEST_F(EnergyPlusFixture, SingZoneCoolHeatSetPtMgrZoneInletNodeTest)
         "   **  Fatal  ** InitSetPointManagers: Errors found in getting SetPointManager input.",
         "   ...Summary of Errors that led to program termination:",
         "   ..... Reference severe error count=2",
-        "   ..... Last severe error=SetpointManager:SingleZone:Cooling=\"COOLING SUPPLY AIR TEMP MANAGER 1\", The zone inlet node of ZNF1 INLET NODE",
+        "   ..... Last severe error=SetpointManager:SingleZone:Cooling=\"COOLING SUPPLY AIR TEMP MANAGER 1\", The zone inlet node of ZNF1 INLET "
+        "NODE",
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/SetPointManager.unit.cc:1594]: FWIW, some of these error snippets that have to match _exactly_ could be modified to use [`compare_err_stream_substring`](https://github.com/NREL/EnergyPlus/blob/92ba541ca525940013c228b84b7644d31212118f/tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.hh#L193).  That way you can just make sure it has a reasonable message, and not have these full text blocks that have to match exactly.

I know you didn't do a functional change here, it just line wrapped, but I'm saying it anyway :)

**Review by Myoldmopar — APPROVED** [review]
This seems quite fine to me.

**Comment by Myoldmopar** [comment]
I think we'll get this merged now, thanks @jmarrec

**Pull request merged**